File: go/pkg/mechbase/zones.go
package mechbase

import (
	"context"
	"fmt"
	"net/http"
	"net/url"
)

// Zones is the zone registry resource.
type Zones struct {
	client *Client
	iid    int
}

// ZoneInput is the create/upsert/update body for zones.
type ZoneInput struct {
	Name              *string `json:"name,omitempty"`
	ExternalID        *string `json:"external_id,omitempty"`
	SectionID         *int    `json:"section_id,omitempty"`
	SectionExternalID *string `json:"section_external_id,omitempty"`
}

// ListZonesOptions filters Zones.List. Zero values are omitted.
type ListZonesOptions struct {
	Q         string
	SectionID int
	Limit     int
	Offset    int
}

// List returns a page of zones.
func (z *Zones) List(ctx context.Context, opts ListZonesOptions) ([]Zone, error) {
	q := url.Values{}
	if opts.Q != "" {
		q.Set("q", opts.Q)
	}
	if opts.SectionID != 0 {
		q.Set("section_id", fmt.Sprintf("%d", opts.SectionID))
	}
	addPaging(q, opts.Limit, opts.Offset)
	var env listEnvelope[Zone]
	if err := z.client.doJSON(ctx, http.MethodGet, pathf(z.iid, "/zones"), q, nil, &env); err != nil {
		return nil, err
	}
	return env.Items, nil
}

// Get fetches one zone by numeric id.
func (z *Zones) Get(ctx context.Context, zoneID int) (*Zone, error) {
	var out Zone
	if err := z.client.doJSON(ctx, http.MethodGet, pathf(z.iid, fmt.Sprintf("/zones/%d", zoneID)), nil, nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Create creates a zone.
func (z *Zones) Create(ctx context.Context, in ZoneInput) (*Zone, error) {
	var out Zone
	if err := z.client.doJSON(ctx, http.MethodPost, pathf(z.iid, "/zones"), nil, in, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Upsert creates or updates a zone, matched by external_id.
func (z *Zones) Upsert(ctx context.Context, in ZoneInput) (*Zone, error) {
	var out Zone
	if err := z.client.doJSON(ctx, http.MethodPut, pathf(z.iid, "/zones"), nil, in, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Update partially updates a zone by numeric id.
func (z *Zones) Update(ctx context.Context, zoneID int, in ZoneInput) (*Zone, error) {
	var out Zone
	if err := z.client.doJSON(ctx, http.MethodPatch, pathf(z.iid, fmt.Sprintf("/zones/%d", zoneID)), nil, in, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Delete removes a zone; cascade also removes children.
func (z *Zones) Delete(ctx context.Context, zoneID int, cascade bool) (*DeleteResult, error) {
	q := url.Values{}
	if cascade {
		q.Set("cascade", "true")
	}
	var out DeleteResult
	if err := z.client.doJSON(ctx, http.MethodDelete, pathf(z.iid, fmt.Sprintf("/zones/%d", zoneID)), q, nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

File: swift/Sources/Mechbase/Zones.swift
import Foundation

/// Create/upsert/update body for zones. Nil fields are omitted by the encoder.
public struct ZoneInput: Encodable, Sendable {
    public var name: String?
    public var externalId: String?
    public var sectionId: Int?
    public var sectionExternalId: String?

    public init(name: String? = nil, externalId: String? = nil,
                sectionId: Int? = nil, sectionExternalId: String? = nil) {
        self.name = name
        self.externalId = externalId
        self.sectionId = sectionId
        self.sectionExternalId = sectionExternalId
    }

    enum CodingKeys: String, CodingKey {
        case name
        case externalId = "external_id"
        case sectionId = "section_id"
        case sectionExternalId = "section_external_id"
    }
}

public struct Zones: Sendable {
    let scope: InstallationScope

    init(http: HTTPClient, installationId: Int) {
        self.scope = InstallationScope(http: http, installationId: installationId)
    }

    public func list(query: String? = nil, sectionId: Int? = nil, limit: Int = 50, offset: Int = 0) async throws -> [Zone] {
        let q: [URLQueryItem] = [
            URLQueryItem(name: "q", value: query),
            URLQueryItem(name: "section_id", value: sectionId.map(String.init)),
            URLQueryItem(name: "limit", value: String(limit)),
            URLQueryItem(name: "offset", value: String(offset)),
        ]
        return try await scope.http.get(scope.path("/zones"), query: q, as: Page<Zone>.self).items
    }

    public func get(id: Int) async throws -> Zone {
        try await scope.http.get(scope.path("/zones/\(id)"), as: Zone.self)
    }

    public func create(_ input: ZoneInput) async throws -> Zone {
        try await scope.http.postJSON(scope.path("/zones"), body: input, as: Zone.self)
    }

    public func upsert(_ input: ZoneInput) async throws -> Zone {
        try await scope.http.putJSON(scope.path("/zones"), body: input, as: Zone.self)
    }

    public func update(id: Int, _ input: ZoneInput) async throws -> Zone {
        try await scope.http.patchJSON(scope.path("/zones/\(id)"), body: input, as: Zone.self)
    }

    public func delete(id: Int, cascade: Bool = false) async throws -> DeleteResult {
        let q = cascade ? [URLQueryItem(name: "cascade", value: "true")] : []
        return try await scope.http.delete(scope.path("/zones/\(id)"), query: q, as: DeleteResult.self)
    }
}
